modules/basegl/src/processors/instancerenderer.cpp

```diff
@@ -0,0 +1,394 @@
+                                   normalize(normal), normalize(toCameraDir));
+    FragData0 = fragColor;
+    PickingData = picking;
+}
+)");
+
+}  // namespace
+
+namespace detail {
+
+DynPortManager::DynPortManager(InstanceRenderer* theRenderer, std::unique_ptr<Inport> aPort,
+                               std::function<size_t()> aSize,
+                               std::function<void(Shader&, size_t)> aSet,
+                               std::function<void(ShaderObject&)> aAddUniform)
+    : renderer{theRenderer}
+    , port{std::move(aPort)}
+    , size{aSize}
+    , set{aSet}
+    , addUniform{aAddUniform} {
+
+    renderer->addPort(*port);
+}
+
+DynPortManager::DynPortManager(DynPortManager&& rhs)
+    : renderer{nullptr}
+    , port{nullptr}
+    , size{std::move(rhs.size)}
+    , set{std::move(rhs.set)}
+    , addUniform{std::move(rhs.addUniform)} {
+    std::swap(rhs.renderer, renderer);
+    std::swap(rhs.port, port);
+}
+
+DynPortManager& DynPortManager::operator=(DynPortManager&& that) {
+    if (this != &that) {
+        std::swap(that.renderer, renderer);
+        std::swap(that.port, port);
+        std::swap(that.size, size);
+        std::swap(that.set, set);
+        std::swap(that.addUniform, addUniform);
+    }
+    return *this;
+}
+DynPortManager::~DynPortManager() {
+    if (port) renderer->removePort(port.get());
+}
+
+template <typename T>
+inline constexpr std::string_view glslTypeName() {
+    using types =
+        std::tuple<float, double, bool, uint32_t, int32_t, vec2, dvec2, bvec2, ivec2, uvec2, vec3,
+                   dvec3, bvec3, ivec3, uvec3, vec4, dvec4, bvec4, ivec4, uvec4>;
+    constexpr std::array<std::string_view, 20> names = {
+        "float", "double", "bool",  "uint",  "int",   "vec2", "dvec2", "bvec2", "ivec2", "uvec2",
+        "vec3",  "dvec3",  "bvec3", "ivec3", "uvec3", "vec4", "dvec4", "bvec4", "ivec4", "uvec4"};
+
+    return names[util::index_of<T, types>()];
+}
+
+template <typename T>
+DynPortManager createDynPortManager(InstanceRenderer* theRenderer, std::string_view identifier,
+                                    StringProperty* uniform) {
+
+    auto port = std::make_unique<DataInport<std::vector<T>>>(identifier);
+    port->setOptional(true);
+    auto size = [p = port.get()]() { return p->hasData() ? p->getData()->size() : size_t(-1); };
+    auto set = [p = port.get(), u = uniform](Shader& shader, size_t index) {
+        if (p->hasData()) {
+            shader.setUniform(u->get(), (*p->getData())[index]);
+        }
+    };
+    auto addUniform = [u = uniform](ShaderObject& so) {
+        so.addSegment(ShaderSegment{
+            irplaceholder::uniform, u->get(),
+            fmt::format("uniform {0} {1} = {0}(0);", detail::glslTypeName<T>(), u->get())});
+    };
+
+    return DynPortManager{theRenderer, std::move(port), std::move(size), std::move(set),
+                          std::move(addUniform)};
+}
+
+}  // namespace detail
+
+InstanceRenderer::InstanceRenderer()
```

Comment:
Most of this would also be valid for a general renderer processor, where one can create custom shaders?

Reply:
maybe?